mypy/subtypes.py

```diff
@@ -143,10 +143,15 @@ def visit_instance(self, left: Instance) -> bool:
             rname = right.type.fullname()
             if not left.type.has_base(rname) and rname != 'builtins.object':
                 return False
-
             # Map left type to corresponding right instances.
             t = map_instance_to_supertype(left, right.type)
-
+            # Check that one of the types does not have type args or
+            # the number of type args is the same and
+            # each left type arg is acceptable in place of the matching right one
+            if not t.args or not right.args:
+                return True
+            if len(t.args) != len(right.args):
+                return False
```

Comment:
I think the bug happens even before. At this stage all the type arguments should have correct counts. For example, if you look at third pass in ``typeanal.py``, then you will find code that puts ``[AnyType()]*whatever_is_needed``. But maybe mypy code itself does not always create ``Instance``s cleanly. For example, ``builtins.tuple`` is not something valid, it should be ``builtins.tuple[Any]``.

> (what does the * on the RHS mean btw?)

``*`` indicates this type is a result of type variable substitution. Typically this happens as a result of type inference.

> but merely the type args don't match

We don't have variadic generic classes, so logically this should never happen.

I think, ideally, you could try to find where those strange ``Instance``s are created (I suppose there are only few such places) and fix them to return ``Instance``s with correct number of type variables.